File: backend/app/models/album.py
from __future__ import annotations

from typing import Optional
from pydantic import BaseModel


class AlbumRelocateRequest(BaseModel):
    destination: str  # absolute path to new parent directory


class AlbumUpdateRequest(BaseModel):
    """Partial update model for PATCH /api/albums/{id}. All fields optional."""
    album: Optional[str] = None
    albumartist: Optional[str] = None
    year: Optional[int] = None
    month: Optional[int] = None
    day: Optional[int] = None
    genres: Optional[str] = None
    label: Optional[str] = None
    country: Optional[str] = None
    albumtype: Optional[str] = None
    albumstatus: Optional[str] = None
    albumdisambig: Optional[str] = None
    comp: Optional[int] = None
    mb_albumid: Optional[str] = None
    mb_releasegroupid: Optional[str] = None
    mb_albumartistid: Optional[str] = None
    catalognum: Optional[str] = None
    barcode: Optional[str] = None
    asin: Optional[str] = None
    original_year: Optional[int] = None


class TrackInAlbum(BaseModel):
    """Lightweight track representation used inside album detail view."""

    id: int
    title: Optional[str] = None
    artist: Optional[str] = None
    track: Optional[int] = None
    disc: Optional[int] = None
    length: Optional[float] = None
    format: Optional[str] = None
    bitrate: Optional[int] = None
    samplerate: Optional[int] = None
    bitdepth: Optional[int] = None
    added: Optional[float] = None
    path: Optional[str] = None


class AlbumSummary(BaseModel):
    """Album representation for grid/list browsing."""

    id: int
    album: Optional[str] = None
    albumartist: Optional[str] = None
    year: Optional[int] = None
    genres: Optional[str] = None
    label: Optional[str] = None
    country: Optional[str] = None
    albumtype: Optional[str] = None
    track_count: int = 0
    total_length: Optional[float] = None
    has_art: bool = False
    added: Optional[float] = None
    # Derived: most common format among items in the album
    format: Optional[str] = None


class AlbumDetail(AlbumSummary):
    """Full album with all metadata fields and track list."""

    month: Optional[int] = None
    day: Optional[int] = None
    disctotal: Optional[int] = None
    comp: Optional[int] = None
    mb_albumid: Optional[str] = None
    mb_releasegroupid: Optional[str] = None
    catalognum: Optional[str] = None
    barcode: Optional[str] = None
    asin: Optional[str] = None
    albumstatus: Optional[str] = None
    albumdisambig: Optional[str] = None
    original_year: Optional[int] = None
    rg_album_gain: Optional[float] = None
    rg_album_peak: Optional[float] = None
    r128_album_gain: Optional[float] = None
    items: list[TrackInAlbum] = []

File: frontend/src/types/album.ts
export interface TrackInAlbum {
  id: number
  title: string
  artist: string
  track: number
  disc: number
  length: number       // seconds
  format: string
  bitrate: number
  samplerate: number
  bitdepth: number
  added: number        // unix timestamp
  path: string | null
}

export interface AlbumSummary {
  id: number
  album: string
  albumartist: string
  year: number
  genres: string | null
  label: string | null
  country: string | null
  albumtype: string | null
  track_count: number
  total_length: number  // seconds
  has_art: boolean
  added: number
  format: string | null
}

export interface AlbumDetail extends AlbumSummary {
  month: number | null
  day: number | null
  disctotal: number
  comp: boolean
  mb_albumid: string | null
  mb_releasegroupid: string | null
  catalognum: string | null
  barcode: string | null
  asin: string | null
  albumstatus: string | null
  albumdisambig: string | null
  original_year: number | null
  rg_album_gain: number | null
  rg_album_peak: number | null
  r128_album_gain: number | null
  items: TrackInAlbum[]
}

export interface PaginatedAlbums {
  items: AlbumSummary[]
  total: number
  page: number
  page_size: number
}
